random.go

```diff
@@ -0,0 +1,142 @@
+package mybench
+
+import (
+	"errors"
+	"math"
+	"math/rand"
+	"sort"
+	"time"
+)
+
+type Rand struct {
+	*rand.Rand
+}
+
+// Creates a new Rand object
+func NewRand() *Rand {
+	return &Rand{
+		Rand: rand.New(rand.NewSource(time.Now().UnixMicro())),
+	}
+}
+
+func (r *Rand) UniformFloat(min, max float64) float64 {
+	return r.Rand.Float64()*(max-min) + min
+}
+
+func (r *Rand) UniformInt(min, max int64) int64 {
+	return r.Rand.Int63n(max-min) + min
+}
+
+func (r *Rand) NormalFloat(mean, stddev float64) float64 {
+	return r.Rand.NormFloat64()*stddev + mean
+}
+
+func (r *Rand) NormalInt(mean, stddev int64) int64 {
+	return int64(r.NormalFloat(float64(mean), float64(stddev)))
+}
+
+func (r *Rand) HistFloat(hist HistogramDistribution) float64 {
+	x := r.Rand.Float64()
+
+	v, err := interp1d(hist.cumulativeDistribution, hist.binsEndPoints, x)
+	if err != nil {
+		panic(err)
+	}
+
+	return v
+}
+
+func (r *Rand) HistInt(hist HistogramDistribution) int64 {
+	return int64(math.Round(r.HistFloat(hist)))
+}
+
+func interpolate(x0, y0, x1, y1, x float64) float64 {
+	return y0 + (x-x0)*(y1-y0)/(x1-x0)
+}
+
+// xs and ys must be sorted and must have the same size
+func interp1d(xs, ys []float64, x float64) (float64, error) {
```

Comment:
The xs and ys terminologies are coming from https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.interp1d.html. Not sure what f and fp are? I'm not sure what you mean by discrete function. This is simply doing a 1D interpolation between points specified in the xs and ys array at x.

Reply:
Its a fancy way of saying f, and fp are two columns in a table:

| foo |  bar |
| --- | --- |
| 3 | 10 |
| 7 | 20 |
| 11 | 30 |
| 17 | 40 |

You define a discrete function f() which maps foo to bar so:

f(3) = 10
..
f(17) = 40

Since the set is discrete we can do the same for the inverse fp:

fp(10) = 3
...
fp(40) = 17

Not sure modeling after scipy matters.

Take it or leave it.  nbd.